src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs

```diff
@@ -421,31 +421,41 @@ protected override bool ReplacementChangesSemanticsForNodeLanguageSpecific(Synta
             {
-                var newSwitchStatement = (SwitchStatementSyntax)currentReplacedNode;
+                // Switch statement's expression changed, verify that the conversions from switch case labels to new switch
+                // expression type are not broken.
 
-                if (originalSwitchStatement.Expression == previousOriginalNode)
-                {
-                    // Switch expression changed, verify that the conversions from switch case labels to new switch expression type are not broken.
+                var newSwitchStatement = (SwitchStatementSyntax)currentReplacedNode;
 
-                    var originalExpressionType = this.OriginalSemanticModel.GetTypeInfo(originalSwitchStatement.Expression, this.CancellationToken).Type;
-                    var newExpressionType = this.SpeculativeSemanticModel.GetTypeInfo(newSwitchStatement.Expression, this.CancellationToken).Type;
+                var originalSwitchLabels = originalSwitchStatement.Sections.SelectMany(section => section.Labels).ToArray();
+                var newSwitchLabels = newSwitchStatement.Sections.SelectMany(section => section.Labels).ToArray();
 
-                    var originalSwitchLabels = originalSwitchStatement.Sections.SelectMany(section => section.Labels).ToArray();
-                    var newSwitchLabels = newSwitchStatement.Sections.SelectMany(section => section.Labels).ToArray();
-                    for (var i = 0; i < originalSwitchLabels.Length; i++)
+                for (var i = 0; i < originalSwitchLabels.Length; i++)
+                {
+                    if (originalSwitchLabels[i] is CaseSwitchLabelSyntax originalSwitchLabel &&
+                        newSwitchLabels[i] is CaseSwitchLabelSyntax newSwitchLabel &&
+                        !ImplicitConversionsAreCompatible(originalSwitchLabel.Value, newSwitchLabel.Value))
                     {
-                        if (originalSwitchLabels[i] is CaseSwitchLabelSyntax originalSwitchLabel)
-                        {
-                            if (newSwitchLabels[i] is CaseSwitchLabelSyntax newSwitchLabel && !ImplicitConversionsAreCompatible(originalSwitchLabel.Value, newSwitchLabel.Value))
-                            {
-                                return true;
-                            }
-                        }
+                        return true;
                     }
                 }
             }
+            else if (currentOriginalNode.IsKind(SyntaxKind.SwitchExpression, out SwitchExpressionSyntax originalSwitchExpression) &&
+                     originalSwitchExpression.GoverningExpression == previousOriginalNode)
+            {
+                var replacedSwitchExpression = (SwitchExpressionSyntax)currentReplacedNode;
+
+                // Switch expression's expression changed.  Ensure it's the same type as before. If not, inference of
+                // the meaning of the patterns within can change.
```

Comment:
❔ Do we not need to check the patterns, like we do for switch statements?

Reply:
it's quite possble we need to for total correctness.  i'm not certain at this point.  going to leave it alone for now :)